File: frontend/src/types/remediation.ts
// Mirrors backend/schemas/remediation.py.

export type ExecutionPath = 'auto_fix' | 'user_approved' | 'human_escalation'

export type ExecutionStatus =
  | 'not_started'
  | 'pending_approval'
  | 'in_progress'
  | 'completed'
  | 'failed'
  | 'escalated'
  | 'rejected'

export type MCPToolStatus = 'success' | 'failed' | 'skipped'

export type ApprovalType = 'auto' | 'user_approved' | 'escalated'

export type PolicyCompliance = 'compliant' | 'violation_overridden' | 'exception'

export type VerificationResult = 'passed' | 'failed' | 'skipped'

/** A single simulated MCP tool invocation within a runbook. */
export interface MCPToolExecution {
  tool: string
  category: string
  input: Record<string, unknown>
  output: Record<string, unknown>
  duration_ms: number
  status: MCPToolStatus
}

/** The safety router's rationale for the chosen execution path. */
export interface SafetyDecision {
  why_safe: string
  approval_required: boolean
  rollback_available: boolean
}

/** Compliance metadata attached to a remediation result. */
export interface AuditCompliance {
  approval_type: ApprovalType
  policy_compliance: PolicyCompliance
  rollback_available: boolean
  retention_expires: string
  persistent_data_modified: boolean
}

/** Module 3 output: full remediation record + report. */
export interface RemediationResult {
  remediation_id: string
  recommendation_id: string
  issue_id: string
  workload_id: string
  execution_path: ExecutionPath
  execution_status: ExecutionStatus
  action_taken: Record<string, unknown>
  reason_for_action: string
  safety_decision: SafetyDecision
  ai_decision_steps: Array<Record<string, unknown>>
  mcp_tools_executed: MCPToolExecution[]
  impact_result: Record<string, unknown>
  execution_timeline: Array<Record<string, unknown>>
  audit_compliance: AuditCompliance
  user_facing_report: string
  rollback_triggered: boolean
  verification_result: VerificationResult
}

File: backend/schemas/remediation.py
"""Pydantic schemas for RemediationResult and its execution structures."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

ExecutionPath = Literal["auto_fix", "user_approved", "human_escalation"]
ExecutionStatus = Literal[
    "not_started",
    "pending_approval",
    "in_progress",
    "completed",
    "failed",
    "escalated",
    "rejected",
]
MCPToolStatus = Literal["success", "failed", "skipped"]
ApprovalType = Literal["auto", "user_approved", "escalated"]
PolicyCompliance = Literal["compliant", "violation_overridden", "exception"]
VerificationResult = Literal["passed", "failed", "skipped"]


class MCPToolExecution(BaseModel):
    """A single simulated MCP tool invocation within a runbook."""

    tool: str
    category: str
    input: dict
    output: dict
    duration_ms: int
    status: MCPToolStatus


class SafetyDecision(BaseModel):
    """The safety router's rationale for the chosen execution path."""

    why_safe: str
    approval_required: bool
    rollback_available: bool


class AuditCompliance(BaseModel):
    """Compliance metadata attached to a remediation result."""

    approval_type: ApprovalType
    policy_compliance: PolicyCompliance
    rollback_available: bool
    retention_expires: datetime
    persistent_data_modified: bool


class RemediationResult(BaseModel):
    """Module 3 output: full remediation record + report."""

    remediation_id: str
    recommendation_id: str
    issue_id: str
    workload_id: str
    execution_path: ExecutionPath
    execution_status: ExecutionStatus
    action_taken: dict
    reason_for_action: str
    safety_decision: SafetyDecision
    ai_decision_steps: list[dict]
    mcp_tools_executed: list[MCPToolExecution]
    impact_result: dict  # before / after / simulated_savings
    execution_timeline: list[dict]
    audit_compliance: AuditCompliance
    user_facing_report: str
    rollback_triggered: bool
    verification_result: VerificationResult
